null pointer dereference when building

```

[ 43%] Building CXX object src/emulator/modules/CMakeFiles/modules.dir/SceAVConfig/SceAVConfig.cpp.o
In file included from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h:20:0,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/module.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.cpp:18:
/home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h: In instantiation of ‘ImportFn bridge(Ret (*)(HostState&, SceUID, const char*, Args ...), const char*) [with Ret = int; Args = {}; ImportFn = std::function<void(HostState&, int)>; SceUID = int]’:
/home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.cpp:156:1:   required from here
/home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h:50:96:   in constexpr expansion of ‘lay_out<>()’
/home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/lay_out_args.h:108:33: error: dereferencing a null pointer
     add_args_to_layout<Args...>(*layout.data(), state);
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/functional:58:0,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/mem/include/mem/mem.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/mem/include/mem/ptr.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge_types.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/read_arg.h:21,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h:21,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/module.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.cpp:18:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/std_function.h:685:7: error: ‘std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = bridge(Ret (*)(HostState&, SceUID, const char*, Args ...), const char*) [with Ret = int; Args = {}; ImportFn = std::function<void(HostState&, int)>; SceUID = int]::<lambda(HostState&, SceUID)>; <template-parameter-2-2> = void; <template-parameter-2-3> = void; _Res = void; _ArgTypes = {HostState&, int}]’, declared using local type ‘bridge(Ret (*)(HostState&, SceUID, const char*, Args ...), const char*) [with Ret = int; Args = {}; ImportFn = std::function<void(HostState&, int)>; SceUID = int]::<lambda(HostState&, SceUID)>’, is used but never defined [-fpermissive]
       function<_Res(_ArgTypes...)>::
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [src/emulator/modules/CMakeFiles/modules.dir/build.make:63: src/emulator/modules/CMakeFiles/modules.dir/SceAVConfig/SceAVConfig.cpp.o] Error 1

```

i've been seing this error for the last few weeks.

It seems the issue is still there

```
[ 43%] Building CXX object src/emulator/modules/CMakeFiles/modules.dir/SceAVConfig/SceAVConfig.cpp.o
In file included from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h:20:0,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/module.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.cpp:18:
/home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h: In instantiation of ‘ImportFn bridge(Ret (*)(HostState&, SceUID, const char*, Args ...), const char*) [with Ret = int; Args = {}; ImportFn = std::function<void(HostState&, CPUState&, int)>; SceUID = int]’:
/home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.cpp:156:1:   required from here
/home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h:44:96:   in constexpr expansion of ‘lay_out<>()’
/home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/lay_out_args.h:108:33: error: dereferencing a null pointer
     add_args_to_layout<Args...>(*layout.data(), state);
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/functional:58:0,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/mem/include/mem/mem.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/mem/include/mem/ptr.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge_types.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/read_arg.h:21,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/bridge.h:21,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/module/include/module/module.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.h:20,
                 from /home/yoshi/src/console/vita/Vita3K/src/emulator/modules/SceAVConfig/SceAVConfig.cpp:18:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/std_function.h:685:7: error: ‘std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = bridge(Ret (*)(HostState&, SceUID, const char*, Args ...), const char*) [with Ret = int; Args = {}; ImportFn = std::function<void(HostState&, CPUState&, int)>; SceUID = int]::<lambda(HostState&, CPUState&, SceUID)>; <template-parameter-2-2> = void; <template-parameter-2-3> = void; _Res = void; _ArgTypes = {HostState&, CPUState&, int}]’, declared using local type ‘bridge(Ret (*)(HostState&, SceUID, const char*, Args ...), const char*) [with Ret = int; Args = {}; ImportFn = std::function<void(HostState&, CPUState&, int)>; SceUID = int]::<lambda(HostState&, CPUState&, SceUID)>’, is used but never defined [-fpermissive]
       function<_Res(_ArgTypes...)>::
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [src/emulator/modules/CMakeFiles/modules.dir/build.make:63: src/emulator/modules/CMakeFiles/modules.dir/SceAVConfig/SceAVConfig.cpp.o] Błąd 1
make[1]: *** [CMakeFiles/Makefile2:1840: src/emulator/modules/CMakeFiles/modules.dir/all] Błąd 2
make: *** [Makefile:152: all] Error 2
```

--- a/src/emulator/module/include/module/lay_out_args.h
+++ b/src/emulator/module/include/module/lay_out_args.h
@@ -18,6 +18,7 @@
 #pragma once
 
 #include "args_layout.h"
+
 #include <tuple>
 
 struct LayoutArgsState {
@@ -98,7 +99,7 @@ constexpr void add_args_to_layout(ArgLayout &head, LayoutArgsState &state) {
     state = std::get<1>(result);
 
     // Recursively add the remaining arguments.
-    if constexpr (sizeof...(Tail) > 0) {
+    if (sizeof...(Tail) > 0) {
         add_args_to_layout<Tail...>(*(&head + 1), state);
     }
 }
@@ -106,7 +107,7 @@ constexpr void add_args_to_layout(ArgLayout &head, LayoutArgsState &state) {
 template <typename... Args>
 constexpr ArgsLayout<Args...> lay_out() {
     ArgsLayout<Args...> layout = {};
-    if constexpr (sizeof...(Args) > 0) {
+    if (sizeof...(Args) > 0) {
         LayoutArgsState state = {};
         add_args_to_layout<Args...>(*layout.data(), state);
     }

--- a/src/emulator/module/include/module/lay_out_args.h
+++ b/src/emulator/module/include/module/lay_out_args.h
@@ -98,14 +98,18 @@ constexpr void add_args_to_layout(ArgLayout &head, LayoutArgsState &state) {
     state = std::get<1>(result);
 
     // Recursively add the remaining arguments.
-    add_args_to_layout<Tail...>(*(&head + 1), state);
+    if constexpr (sizeof...(Tail) > 0) {
+        add_args_to_layout<Tail...>(*(&head + 1), state);
+    }
 }
 
 template <typename... Args>
 constexpr ArgsLayout<Args...> lay_out() {
     ArgsLayout<Args...> layout = {};
-    LayoutArgsState state = {};
-    add_args_to_layout<Args...>(*layout.data(), state);
+    if constexpr (sizeof...(Args) > 0) {
+        LayoutArgsState state = {};
+        add_args_to_layout<Args...>(*layout.data(), state);
+    }
 
     return layout;
 }

